### ISSUE TITLE
How to get confidence score for recommended packages?

### ISSUE THREAD

**Issue by abs51295** [david-cortes/hpfrec#4]
Hey @david-cortes , 

I can see the package numbers in recommendation. Along with that, can I see the confidence score of the recommended packages as well? Based on my understanding of the code so far, I think it's the same score that's used to sort the packages in `topN()` function. But, I am unable to figure out as to how can I get that score back. I guess it will be a nice metric to know. Let me know if you have any ideas and if I can send a  PR to include that in the library. Thanks.

**Comment by david-cortes** [comment]
If you mean to say “confidence score” as in “the parameter of the Poisson distribution predicted for a given user and item”, then you can use the function `predict` for that. If you mean “confidence score” as in some measurement that includes variance, that’s not implemented in the package but you could nevertheless construct a function around it that would sample gamma parameters.

If you want to get predicted values along with topN, you’d have to slice the “allpreds” array with the same values that are returned (before reindexing if using `reindex=True`), and multiply that result by `-1`. Be aware that there are multiple return sections depending on the parameters (reindexing, items_pool, exclude_seen).

**Comment by abs51295** [comment]
Well, `“the parameter of the Poisson distribution predicted for a given user and item”` will be greater than 1 in some cases. What exactly I am looking for is the probability for the predicted items. Do you have any ideas as to how to calculate that?

**Comment by david-cortes** [comment]
Well, in that case you’re very out of luck because this is a Poisson model that assumes count data, not binary data. You could calculate the probability that some user-item combination would be greater than zero according to this model (given by `1 – exp(-prediction)`, but I’m not sure how well that would work out for binary data and since the model is fit to all user-item combinations, they will be very, very small numbers for any combination that was not non-zero in the training data.

You won’t be able to get such numbers from implicit-ALS and related either, as those, while taking binary entries, minimize square loss (thus not probabilistic) – you’d have to fit some different model based on log-loss (Bernoulli likelihood), but those are orders of magnitude slower as they require iterating over all the entries and not just the non-zero ones (and when they don’t, they wouldn’t reflect real probabilities anyway).

I think your best bet would be to take some sample entries of predicted values from this model and whether the real entry was non-zero, fit a logistic regression or some other calibration model to them, and then apply that calibration model on top of the predicted values from this model.

**Comment by abs51295** [comment]
> Well, in that case you’re very out of luck because this is a Poisson model that assumes count data, not binary data.

How to refactor code so that it can work on binary data as well? I actually see a provision of fitting the model on binary data in the implementation done by the paper authors: https://github.com/premgopalan/hgaprec/blob/master/README#L63-L64


> I think your best bet would be to take some sample entries of predicted values from this model and whether the real entry was non-zero, fit a logistic regression or some other calibration model to them, and then apply that calibration model on top of the predicted values from this model.

So, If I understand correctly, what you are saying here is considering the predicted value as feature and the real entry as non-zero as label (0 for zero, 1 for non-zero) and treat it as a two class classification problem and then calculate the probability of the predicted value belonging to class 1. But again, I am doubtful about how much accurate this will be.

**Comment by david-cortes** [comment]
I’m not familiar with the author’s original code, but be aware that it contains more than just the model described in the paper (e.g. it offers a non-hierarchical model too), while this repo implements only the  HPF model:
https://arxiv.org/pdf/1311.1704

From a quick look at his code, it seems that the binary option will only take the entries as zero or one and fit a Poisson model on top of it, even though they are binary (at least if you choose the HPF version in the other parameters), but don't take my word for it.

So yes, if I were you I’d go for the recalibration approach (logistic regression with bias + prediction as feature).

**Comment by abs51295** [comment]
> From a quick look at his code, it seems that the binary option will only take the entries as zero or one and fit a Poisson model on top of it, even though they are binary (at least if you choose the HPF version in the other parameters), but don't take my word for it.

Well, how much difference will it make for HPF to be fitted on count data vs binary data? What I am curious about is we can treat binary data as count data as well by keeping the count values to 1. Will the predicted values be 0 or 1 in the case of binary data?

**Comment by david-cortes** [comment]
I’m not sure how well it would work on binary data compared to other models that are meant for binary data (I guess for very sparse inputs it should be competitive); and if your data is not binary, you should get worse results from binarizing it.

Even if you do binarize it and fit this model on top, it’s still a Poisson model and the predicted values are not guaranteed to be below 1 (in most likelihood though, they will all be, but are still by no means reflective of probabilities).

The model itself does not have any easy adaptation for probabilistic binary outputs, you would have to change the model completely – for example, you can take this one as a base and see how the results compare to HPF:
https://github.com/david-cortes/binmf
It uses hinge loss, thus it’s not probabilistic either (and it sub-samples entries with zero value rather than using them all), but the results should be similar to a Bernoulli (log-loss) model (the update rules for log-loss are rather similar to the algorithm used there).

**Comment by abs51295** [comment]
Cool @david-cortes. Thanks for the detailed explanation. I will try out the logistic regression approach and let you know. But again, the values from the `predict()` function look very close to each other for predicted items. I don't know if that's intended or I did something wrong while fitting the model. But, in that case I am not sure if the logistic regression model will work as expected. Also, why are the values from the `predict()` function lower for top packages in some cases than the bottom. Aren't topN packages sorted descending by that value?

**Comment by david-cortes** [comment]
The values from `predict` should be relatively close to each other and very small because the model is fit mostly to entries with zero value – this is not a problem as long as it ranks the items properly. Using smaller priors should also make these values fluctuate more, but only up to a certain point.

The `topN` function should indeed sort by the same values that `predict` outputs. If possible, please pickle the model that is producing this unexpected behavior and attach it here.

**Comment by abs51295** [comment]
> The `topN` function should indeed sort by the same values that `predict` outputs. If possible, please pickle the model that is producing this unexpected behavior and attach it here.

Well, Github didn't allow me to attach it here. So, I uploaded it to drive: https://drive.google.com/file/d/1qCkMwofcpUcKUoWGM6oRB4xtAKliaEiI/view?usp=sharing

**Comment by david-cortes** [comment]
> > The `topN` function should indeed sort by the same values that `predict` outputs. If possible, please pickle the model that is producing this unexpected behavior and attach it here.
> 
> Well, Github didn't allow me to attach it here. So, I uploaded it to drive: https://drive.google.com/file/d/1qCkMwofcpUcKUoWGM6oRB4xtAKliaEiI/view?usp=sharing

I've been trying to look for results of `topN` that are not ordered and haven't been able to find any. Could you point me to some user for which the `predict` function would not sort items in the same order as `topN`?

**Comment by abs51295** [comment]
> > > The `topN` function should indeed sort by the same values that `predict` outputs. If possible, please pickle the model that is producing this unexpected behavior and attach it here.
> > 
> > 
> > Well, Github didn't allow me to attach it here. So, I uploaded it to drive: https://drive.google.com/file/d/1qCkMwofcpUcKUoWGM6oRB4xtAKliaEiI/view?usp=sharing
> 
> I've been trying to look for results of `topN` that are not ordered and haven't been able to find any. Could you point me to some user for which the `predict` function would not sort items in the same order as `topN`?

Oops, my bad. I was converting the output to a set and hence the output was unordered. Sorry for the misunderstanding.
